lib/HTFeed/QueueRunner.pm

```diff
@@ -0,0 +1,149 @@
+package HTFeed::QueueRunner;
+
+use HTFeed::StagingSetup;
+use HTFeed::Bunnies;
+use HTFeed::Config qw(get_config set_config);
+use HTFeed::DBTools qw(update_queue);
+use Log::Log4perl qw(get_logger);
+use HTFeed::Job;
+use Sys::Hostname qw(hostname);
+use File::Path qw(remove_tree);
+
+use strict;
+
+# Gets items to ingest from queue; processes them start to finish.
+
+sub new {
+  my $class = shift;
+  my %params = @_;
+
+  my $self = {};
+  bless($self, $class);
+
+  # TODO: HTFeed::VolumeList that can be used as the queue instead
+  $self->{queue} = $params{queue} || HTFeed::Bunnies->new();
+  $self->{clean} = $params{clean} || get_config('clean');
+  # update status in the database by default
+  $self->{staging_root} = $params{staging_root} || $self->node_stage_dir;
+  $self->{timeout} = $params{timeout};
+  $self->{update_db} = $params{update_db};
+  $self->{update_db} = 1 if not defined $self->{update_db};
+  # for testing
+  $self->{should_fork} = $params{should_fork};
+  $self->{should_fork} = 1 if not defined $self->{should_fork};
+
+  set_config($self->{staging_root}, 'staging_root');
+  HTFeed::StagingSetup::make_stage();
+
+  return $self;
+
+}
+
+sub node_stage_dir {
+  my $self = shift;
+
+  get_config('staging_root') . '/'. hostname;
+}
+
+sub run {
+  my $self = shift;
+
+  $self->setup_signal_handlers;
+
+  print("feedd QueueRunner running, waiting for something to ingest, pid = $$\n");
+
+  while( my $job_info = $self->{queue}->next_job($self->{timeout}) ){
+    $self->marshal_and_run($job_info);
```

Comment:
Should check that the job hasn't been re-delivered too many times already as that might indicate a state where it's just going to crash & fail again.